src/WinRT.Runtime/ObjectReference.cs

```diff
@@ -1,305 +1,334 @@
-            {
-                Guid iid = typeof(TInterface).GUID;
-                Marshal.ThrowExceptionForHR(VftblIUnknown.QueryInterface(ThisPtr, ref iid, out IntPtr comPtr));
-                try
-                {
-                    return (TInterface)Marshal.GetObjectForIUnknown(comPtr);
-                }
-                finally
-                {
-                    var vftblPtr = Unsafe.AsRef<WinRT.VftblPtr>(comPtr.ToPointer());
-                    var vftblIUnknown = Marshal.PtrToStructure<WinRT.Interop.IUnknownVftbl>(vftblPtr.Vftbl);
-                    vftblIUnknown.Release(comPtr);
-                }
-            }
-
-#if NETSTANDARD2_0
-            return (TInterface)typeof(TInterface).GetHelperType().GetConstructor(new[] { typeof(IObjectReference) }).Invoke(new object[] { this });
-#else
-            return (TInterface)(object)new WinRT.IInspectable(this);
-#endif
-        }
-
-        public int TryAs<T>(out ObjectReference<T> objRef) => TryAs(GuidGenerator.GetIID(typeof(T)), out objRef);
-
-        public virtual unsafe int TryAs<T>(Guid iid, out ObjectReference<T> objRef)
-        {
-            objRef = null;
-            ThrowIfDisposed();
-            int hr = VftblIUnknown.QueryInterface(ThisPtr, ref iid, out IntPtr thatPtr);
-            if (hr >= 0)
-            {
-                objRef = ObjectReference<T>.Attach(ref thatPtr); 
-            }
-            return hr;
-        }
-
-        public unsafe IObjectReference As(Guid iid) => As<IUnknownVftbl>(iid);
-
-        public T AsType<T>()
-        {
-            ThrowIfDisposed();
-            var ctor = typeof(T).GetConstructor(new[] { typeof(IObjectReference) });
-            if (ctor != null)
-            {
-                return (T)ctor.Invoke(new[] { this });
-            }
-            throw new InvalidOperationException("Target type is not a projected interface.");
-        }
-
-        public IntPtr GetRef()
-        {
-            ThrowIfDisposed();
-            AddRef();
-            return ThisPtr;
-        }
-
-        protected void ThrowIfDisposed()
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using WinRT.Interop;
+
+#pragma warning disable 0169 // The field 'xxx' is never used
+#pragma warning disable 0649 // Field 'xxx' is never assigned to, and will always have its default value
+
+namespace WinRT
+{
+    public abstract class IObjectReference : IDisposable
+    {
+        protected bool disposed;
+        private readonly IntPtr _thisPtr;
+        private object _disposedLock = new object();
+
+        public IntPtr ThisPtr
+        {
+            get
+            {
+                ThrowIfDisposed();
+                return _thisPtr;
+            }
+        }
+
+#if DEBUG
+        private unsafe uint RefCount 

```

Comment:
I don't believe this needs to be marked `unsafe`.

Reply:
needed for ThisPtr: CS0214: Pointers and fixed size buffers may only be used in an unsafe context